README.md
# heckle

The Clojure(Script) validation library you deserve.

> Heckle is a *flexible*, *function-oriented*, and *composable* validations library for Clojure and ClojureScript.

[![Build Status](https://travis-ci.org/nwallace/heckle.svg)](https://travis-ci.org/nwallace/heckle)
[![Clojars Project](https://img.shields.io/clojars/v/heckle.svg)](https://clojars.org/heckle)

## Why another validation library?

Clojure has several existing validation libraries, but as I looked around I was stunned to see how inflexibly they were designed. Many only support working with hash-maps. Many only support validating one piece of data in isolation (making it hard to validate things like a "Password confirmation" field). I couldn't find a validations library that elegantly addressed all of my validation needs, so I built one.

## Usage

Let's dive right in with a quick example:

```clojure
(require '[heckle.core :refer [validate]]
         '[heckle.validations :as v])

(def login-validations
  [(v/matches #".@." :email)
   (v/matches #"[a-z]" :password "must include a lower-case letter")
   (v/matches #"[A-Z]" :password "must include a capital letter")
   (v/matches #"\d" :password "must include a number")
   (v/length-is-at-least 8 :password)])

(validate login-validations {:email "me@example.com" :password "MyPa55word"})
  ; => {}

(validate login-validations {:email "me" :password "passwd"})
  ; => {:email #{"is invalid"}
  ;     :password #{"must include a capital letter"
  ;                 "must include a number"
  ;                 "must be at least 8 characters"}}
```

As you can see, `heckle.core/validate` take a list of "validations" and some data, and returns a hash-map of errors. The errors is a hash-map where the key is the invalid field and the value is a list of error messages. When everything is okay, the errors hash-map is empty.

All the hard work is handled by the validation functions you provide. The validation functions have a very simple interface: they take the data to validate as their only argument, and they return error information in the form of `[error-key error-message]` if the data fails the validation, or `nil` if the data passes.

The simplicity of the validation functions is the key difference between Heckle and other validation libraries. Since they receive the entire input data, they can check one field or many fields at will. The error key they return is independent from the data its validating.

Heckle ships with many standard validation functions built-in, but it's easy to define your own:

```clojure
(defn can-vote [data]
  (when (< (:age data) 18)
    [:voter "is too young to vote"]))

(defn candidate-is-valid [data]
  (when-not (contains? #{"Clinton" "Stein" "Johnson"} (:candidate data))
    [:candidate "is a menace to society"]))

(heckle.core/validate [can-vote candidate-is-valid] {:age 28 :candidate "Clinton"})
  ; => {}

(heckle.core/validate [can-vote candidate-is-valid] {:age 17 :candidate "Trump"})
  ; => {:voter #{"is too young to vote"}
  ;     :candidate #{"is a menace to society"}}
```

## Built-in validation function builders

Heckle comes with functions to build your own validation functions. The `heckle.validations` namespace contains builder functions for validationg hash-map input data. Example usage: 

```clojure
(heckle.core/validate
  [(heckle.validations/is-present :email)
   (heckle.validations/is-confirmed :email)
   (heckle.validations/is-at-least 18 :age "is too young")]
  {:email "me@example.com" :age 30}) ; => {}
```

| Fn name                  | Description                                                                                       | Arguments                                                                                                                                                                                               | Default error message                    |
| ---                      | ---                                                                                               | ---                                                                                                                                                                                                     | ---                                      |
| `is-present`             | If value is a string, ensures it is not blank. If value is not a string, ensures it is not `nil`. | `key` - the key whose value we will check<br/>`error-msg` - (optional) custom error message when invalid                                                                                                | `"is required"`                          |
| `matches`                | Ensure a string matches against a regular expression                                              | `regex` - the regular expression<br/>`key` - the key whose value we will check<br/>`error-msg` - (optional) custom error message when invalid                                                           | `"is invalid"`                           |
| `is-one-of`              | Ensure a value is one of a list of values                                                         | `collection` - the set of permissible values<br/>`key` - the key whose value we will check<br/>`error-msg` - (optional) custom error message when invalid                                               | e.g. `"must be either A, B or C"`        |
| `is-confirmed`           | Ensure a value has been confirmed accurately (the value and its confirmation are equal)           | `key` - the key whose value we will check is confirmed<br/>`confirmation-key` - (optional) the name of the key of the confirmation value<br/>`error-msg` - (optional) custom error message when invalid | `"does not match"`                       |
| `length-is-at-least`     | Ensure a string or other sequence is at least the given length                                    | `min-length` - the minimum permissible length<br/>`key` - the key whose value we will check<br/>`error-msg` - (optional) custom error message when invalid                                              | e.g. `"must be at least 8 characters"`   |
| `length-is-no-more-than` | Ensure a string or other sequence is at most the given length                                     | `max-length` - the maximum permissible length<br/>`key` - the key whose value we will check<br/>`error-msg` - (optional) custom error message when invalid                                              | e.g. `"must be less than 25 characters"` |
| `is-at-least`            | Ensure a value is greater than or equal to a minimum value                                        | `bound` - the minimum permissible value<br/>`key` - the key whose value we will check<br/>`error-msg` - (optional) custom error message when invalid                                                    | e.g. `"must be at least 3"`              |
| `is-greater-than`        | Ensure a value is strictly greater than a given value                                             | `bound` - the value above which valid values must be<br/>`key` - the key whose value we will check<br/>`error-msg` - (optional) custom error message when invalid                                       | e.g. `"must be greater than 2"`          |
| `is-less-than`           | Ensure a value is strictly less than a given value                                                | `bound` - the value below which valid values must be<br/>`key` - the key whose value we will check<br/>`error-msg` - (optional) custom error message when invalid                                       | e.g. `"must be less than 10"`            |
| `is-no-more-than-than`   | Ensure a value is less than or equal to a maximum value                                           | `bound` - the maximum permissible value<br/>`key` - the key whose value we will check<br/>`error-msg` - (optional) custom error message when invalid                                                    | e.g. `"must be no more than 9"`          |

## Building your own validation functions

If none of the built-in validation functions work for you, it's easy to write your own validation functions. You can just write a your own validation from scratch (it's very easy!), or you can use Heckle's helper functions `make-claim` and `make-denial`.

#### Completely custom validations

A validation is just a function that accepts the input data as an argument and returns `nil` if everything is okay, or error information in the form `[error-key error-message]` if there's an error.

Here are some completely custom validation functions you can use with `validate`:

```clojure
(def sign-up-validations
  [(fn [data] (when (empty? (:email data)) [:email "is required"]))
   (fn [data] (when-not (re-find #"[a-z]" (:password data)) [:password "must contain at least one lowercase letter"]))
   (fn [data] (when-not (re-find #"[A-Z]" (:password data)) [:password "must contain at least one capital letter"]))
   (fn [data] (when-not (re-find #"\d" (:password data)) [:password "must contain at least one number"]))])

(heckle.core/validate sign-up-validations {:email "" :password "pass"})
  ; => {:email #{"is required"}
  ;     :password #{"must conatin at least one capital letter"
  ;                 "must contain at least one number"}}
```

#### Using `make-claim` and `make-denial`

It's even easier to write custom validation functions using two helper functions `make-claim` and `make-denial`. These functions both take a predicate function, the error key and the error message. `make-claim` expects the predicate to return a truthy value when there is no error, and `make-denial` expects the predicate to return a falsey value when there is no error.

Let's see the same example as above, but using these helper functions:

```clojure
(def sign-up-validations
  [(heckle.core/make-denial #(empty? (:email %1)) :email "is required")
   (heckle.core/make-claim #(re-find #"[a-z]" (:password %1)) :password "must contain at least one lowercase letter")
   (heckle.core/make-claim #(re-find #"[A-Z]" (:password %1)) :password "must contain at least one capital letter")
   (heckle.core/make-claim #(re-find #"\d" (:password %1)) :password "must contain at least one number")])

(heckle.core/validate sign-up-validations {:email "" :password "pass"})
  ; => {:email #{"is required"}
  ;     :password #{"must conatin at least one capital letter"
  ;                 "must contain at least one number"}}
```

## Roadmap

* Compile to CLJS as well
* Add collection validations
* Add a syntax shortcut for setting up multiple validations for the same field
* Enable validation to short-circuit other validations for invalid field
* Enable validation to short-circuit all other validations upon first error on any field
* Allow customization of the default error messages
* Support records in addition to hash-maps for built-in validation functions
* Add data type validations?

## License

Copyright © 2018 Nathan Wallace

Distributed under the MIT License.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-03-29
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-03-29
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/heckle/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/heckle/compare/0.1.0...0.1.1


test/heckle/core_test.clj
(ns heckle.core-test
  (:require [midje.sweet :refer :all]
            [heckle.core :refer :all]))

(defn email-presence-validation [data]
  (when (empty? (get data :email "")) [:email "is required"]))
(defn email-format-validation [data]
  (when-not (re-find #".@." (get data :email "")) [:email "is invalid"]))
(defn password-presence-validation [data]
  (when (empty? (get data :password "")) [:password "is required"]))

(facts "about 'validate"
       (fact "it returns no errors when no validations are given"
             (validate [] {}) => {}
             (validate [] {:data "value"}) => {})
       (fact "it returns no errors when the data passes all validations"
             (validate [email-presence-validation]
                       {:email "any non-empty string"}) => {}
             (validate [email-presence-validation
                        email-format-validation]
                       {:email "me@example.com"}) => {}
             (validate [email-presence-validation
                        password-presence-validation]
                       {:email "me@example.com" :password "pass"}) => {})
       (fact "it returns error messages grouped by error key when the data"
             (validate [email-presence-validation]
                       {}) => {:email #{"is required"}}
             (validate [email-presence-validation
                        email-format-validation
                        password-presence-validation]
                       {}) => {:email #{"is required" "is invalid"}
                               :password #{"is required"}}))

(facts "about 'make-claim"
       (let [only-even (make-claim even? :number "must be even")]
         (fact "it returns a function that returns `nil` when called with data that passes the predicate"
               (only-even 2) => nil)
         (fact "it returns a function that returns an error tuple with the given key and message when called with data that fails the predicate"
               (only-even 1) => [:number "must be even"])))

(facts "about 'make-denial"
       (let [only-even (make-denial odd? :number "must be even")]
         (fact "it returns a function that returns `nil` when called with data that passes the predicate"
               (only-even 2) => nil)
         (fact "it returns a function that returns an error tuple with the given key and message when called with data that fails the predicate"
               (only-even 1) => [:number "must be even"])))


src/heckle/core.clj
(ns heckle.core)

(defn validate
  "Check the given validation functions against the given data and return a (possibly empty) collection of errors.

Arguments:
  - `validations` - a list of validation functions to be checked against the data. Each validation function should return a tuple of `[error-key error-msg]` if the validation failed or `nil` otherwise.
  - `data` - the data you're trying to validate (typically a hash-map, though it could be anything else you want as long as the validation functions know how to handle it)

Return value:
  Returns a hash-map of error messages grouped by their key. If all validations pass, it returns an empty hash-map. Otherwise, the hash-map will look something like this: `{key1 #{message1}, key2 #{message2 message3}}`

Example:
  ```
  (def sign-up-validations
    [(fn [data] (when (empty? (:email data)) [:email \"is required\"]))
     (fn [data] (when-not (re-find #\"[a-z]\" (:password data)) [:password \"must contain at least one lowercase letter\"]))
     (fn [data] (when-not (re-find #\"[A-Z]\" (:password data)) [:password \"must contain at least one capital letter\"]))
     (fn [data] (when-not (re-find #\"\\d\" (:password data)) [:password \"must contain at least one number\"]))])

  (heckle.core/validate
    sign-up-validations
    {:email \"\" :password \"pass\"})
  ; => {:email #{\"is required\"}
        :password #{\"must conatin at least one capital letter\"
                    \"must contain at least one number\"}}

  (heckle.core/validate
    sign-up-validations
    {:email \"me@example.com\" :password \"s3cRet\"})
  ; => {}
  ```"
  [validations data]
  (reduce (fn [errors validation]
            (if-let [[error-key error-msg] (validation data)]
              (update errors error-key (fnil conj #{}) error-msg)
              errors))
          {}
          validations))

(defn make-claim
  "A helper function to create a positive validation function.

Arguments:
  - `pred` - a predicate function that should return a truthy value when there is no error
  - `error-key` - will be returned as the first item in a tuple if the predicate check fails
  - `error-msg` - will be returned as the second item in a tuple if the predicate check fails

Return value:
  Returns a validation function that fails if the predicate check fails.

Example:
  ```
  (def email-is-present-validation
    (heckle.core/make-claim
      #(seq (get %1 :email \"\"))
      :email \"is required\"))

  (email-is-present-validation {:email \"\"}) ; => [:email \"is required\"]
  (email-is-present-validation {:email \"me@example.com\"}) ; => nil
  ```"
  [pred error-key error-msg]
  (fn [data]
    (when-not (pred data)
      [error-key error-msg])))

(defn make-denial
  "A helper function to create a negative validation function.

Arguments:
  - `pred` - a predicate function that should return a falsey value when there is no error
  - `error-key` - will be returned as the first item in a tuple if the predicate check fails
  - `error-msg` - will be returned as the second item in a tuple if the predicate check fails

Return value:
  Returns a validation function that fails if the predicate check passes.

Example:
  ```
  (def email-is-present-validation
    (heckle.core/make-denial
      #(empty? (get %1 :email \"\"))
      :email \"is required\"))

  (email-is-present-validation {:email \"\"}) ; => [:email \"is required\"]
  (email-is-present-validation {:email \"me@example.com\"}) ; => nil
  ```"
  [pred error-key error-msg]
  (fn [data]
    (when (pred data)
      [error-key error-msg])))


src/heckle/validations.clj
(ns heckle.validations
  (:require [heckle.core :refer [make-claim make-denial]]
            [clojure.string :as str]))

(defn- to-sentence
  ([coll] (to-sentence coll " and "))
  ([coll conjunction]
   (str/join conjunction
             (if-let [for-commas (seq (drop-last coll))]
               [(str/join ", " for-commas) (last coll)]
               [(last coll)]))))
(defn- pluralize
  [number singular]
  (str number " " singular (when (not= number 1) "s")))

;; general validations
(defn is-present
  "Returns a validator fn that ensures the value of the given field is present (not nil and not a blank string).

Arguments:
  - `key` - the key whose value we will check
  - `error-msg` - (optional) custom error message (defaults to \"is required\")

Return value:
  Returns a validation function that fails if the value of the given field is nil or a blank string.

Example:
  ```
  ((is-present :email) {:email \"me\"}) ; => nil
  ((is-present :email) {:email \"\"}) ; => [:email \"is required\"]
  ((is-present :email \"Uh oh\") {}) ; => [:email \"Uh oh\"]
  ```
"
  ([key] (is-present key "is required"))
  ([key error-msg] (make-claim
                    #(when-let [value (get %1 key "")]
                       (or (not (string? value))
                           (re-find #"\S" value)))
                    key error-msg)))

(defn is-one-of
  "Returns a validator fn that ensures the value of the given field is a member of the given set of permissible values.

Arguments:
  - `collection` - the set of permissible values
  - `key` - the key whose value we will check
  - `error-msg` - (optional) custom error message (defaults to \"must be either `(to-sentence collection)`\")

Return value:
  Returns a validation function that fails if the value of the given field is not a member of `collection`.

Example:
  ```
  ((is-one-of #{:red :blue} :color) {:color :blue}) ; => nil
  ((is-one-of #{:red :blue} :color) {:color :green}) ; => [:color \"must be either :red or :blue\"]
  ((is-one-of #{:red :blue} :color \"Uh oh\") {}) ; => [:color \"Uh oh\"]
  ```
  "
  ([collection key] (is-one-of collection key
                               (str "must be either " (to-sentence collection " or "))))
  ([collection key error-msg] (make-claim
                               #(some #{(get %1 key)} collection)
                               key error-msg)))

(defn is-confirmed
  "Returns a validator fn that ensures the value of the given field matches the value of its confirmation field.

Arguments:
  - `key` - the key whose value we will check is confirmed
  - `confirmation-key` - (optional) the key whose value should match that of `key` (defaults to :`key`-confirmation)
  - `error-msg` - (optional) custom error message (defaults to \"does not match\")

Return value:
  Returns a validation function that fails if the value of the given field doesn't match the value of the confirmation field.

Example:
  ```
  ((is-confirmed :password) {:password \"abcd\" :password-confirmation \"abcd\"}) ; => nil
  ((is-confirmed :password) {:password \"abcd\" :password-confirmation \"bacd\"}) ; => [:password-confirmation \"doesn't match\"]
  ((is-confirmed :password :pswd-confirmation \"Uh oh\") {}) ; => [:pswd-confirmation \"Uh oh\"]
  ```
  "
  ([key] (is-confirmed key (keyword (str (name key) "-confirmation"))))
  ([key confirmation-key] (is-confirmed key confirmation-key "does not match"))
  ([key confirmation-key error-msg]
   (make-claim #(and (contains? %1 key)
                     (contains? %1 confirmation-key)
                     (= (key %1) (confirmation-key %1)))
               confirmation-key error-msg)))

;; string validations
(defn matches
  "Returns a validator fn that ensures the value of the given field matches the given regular expression.

Arguments:
  - `regex` - the regular expression the value should match
  - `key` - the key whose value we will check
  - `error-msg` - (optional) custom error message (defaults to \"is invalid\")

Return value:
  Returns a validation function that fails if the value of the given field doesn't match `regex`.

Example:
  ```
  ((matches #\"\\d\" :password) {:password \"pass123\"}) ; => nil
  ((matches #\"\\d\" :password) {:password \"pass\"}) ; => [:password \"is invalid\"]
  ((matches #\"\\d\" :password \"Uh oh\") {}) ; => [:password \"Uh oh\"]
  ```
  "
  ([regex key] (matches regex key "is invalid"))
  ([regex key error-msg] (make-claim
                          #(when-let [value (get %1 key "")]
                             (re-find regex value))
                          key error-msg)))

(defn length-is-at-least
  "Returns a validator fn that ensures the value of the given field is at least as long as the given length.

Arguments:
  - `min-length` - the minimum permissible length
  - `key` - the key whose value we will check
  - `error-msg` - (optional) custom error message (defaults to \"must be at least `min-length` characters\")

Return value:
  Returns a validation function that fails if the value of the given field is shorter than `min-length`.

Example:
  ```
  ((length-is-at-least 8 :password) {:password \"password\"}) ; => nil
  ((length-is-at-least 8 :password) {:password \"pass\"}) ; => [:password \"must be at least 8 characters\"]
  ((length-is-at-least 8 :password \"Uh oh\") {}) ; => [:password \"Uh oh\"]
  ```
  "
  ([min-length key] (length-is-at-least
                     min-length key
                     (str "must be at least " (pluralize min-length "character"))))
  ([min-length key error-msg] (make-denial
                               #(-> (get %1 key "") count (< min-length))
                               key error-msg)))

(defn length-is-no-more-than
  "Returns a validator fn that ensures the value of the given field is at most as long as the given length.

Arguments:
  - `max-length` - the maximum permissible length
  - `key` - the key whose value we will check
  - `error-msg` - (optional) custom error message (defaults to \"must be no more than `(inc max-length)` characters\")

Return value:
  Returns a validation function that fails if the value of the given field is shorter than `min-length`.

Example:
  ```
  ((length-is-no-more-than 4 :id) {:id \"abcd\"}) ; => nil
  ((length-is-no-more-than 4 :id) {:id \"abcde\"}) ; => [:id \"must be less than 5 characters\"]
  ((length-is-no-more-than 4 :id \"Uh oh\") {}) ; => [:password \"Uh oh\"]
  ```
  "
  ([max-length key] (length-is-no-more-than
                     max-length key
                     (str "must be less than " (pluralize (inc max-length) "character"))))
  ([max-length key error-msg] (make-denial
                               #(-> (get %1 key "") count (> max-length))
                               key error-msg)))

(defn is-at-least
  "Returns a validator fn that ensures the value of the given field is at least the value given.

Arguments:
  - `bound` - the minimum permissible value
  - `key` - the key whose value we will check
  - `error-msg` - (optional) custom error message (defaults to \"must be at least `bound`\")

Return value:
  Returns a validation function that fails if the value of the given field is less than `bound`.

Example:
  ```
  ((is-at-least 18 :age) {:age 18}) ; => nil
  ((is-at-least 18 :age) {:age 17}) ; => [:age \"must be at least 18\"]
  ((is-at-least 18 :age \"Uh oh\") {}) ; => [:age \"Uh oh\"]
  ```
  "
  ([bound key] (is-at-least bound key (str "must be at least " bound)))
  ([bound key error-msg] (make-claim
                          #(and (contains? %1 key)
                                (not (pos? (compare bound (get %1 key)))))
                          key error-msg)))

(defn is-greater-than
  "Returns a validator fn that ensures the value of the given field is strictly greater than the value given.

Arguments:
  - `bound` - the value above which valid values must be
  - `key` - the key whose value we will check
  - `error-msg` - (optional) custom error message (defaults to \"must be greater than `bound`\")

Return value:
  Returns a validation function that fails if the value of the given field is less than or equal to `bound`.

Example:
  ```
  ((is-greater-than 20 :age) {:age 21}) ; => nil
  ((is-greater-than 20 :age) {:age 20}) ; => [:age \"must be greater than 20\"]
  ((is-greater-than 20 :age \"Uh oh\") {}) ; => [:age \"Uh oh\"]
  ```
  "
  ([bound key] (is-greater-than bound key (str "must be greater than " bound)))
  ([bound key error-msg] (make-claim
                          #(and (contains? %1 key)
                                (neg? (compare bound (get %1 key))))
                          key error-msg)))

(defn is-less-than
  "Returns a validator fn that ensures the value of the given field is strictly less than the value given.

Arguments:
  - `bound` - the value below which valid values must be
  - `key` - the key whose value we will check
  - `error-msg` - (optional) custom error message (defaults to \"must be less than `bound`\")

Return value:
  Returns a validation function that fails if the value of the given field is greater than or equal to `bound`.

Example:
  ```
  ((is-less-than 80 :age) {:age 79}) ; => nil
  ((is-less-than 80 :age) {:age 80}) ; => [:age \"must be less than 80\"]
  ((is-less-than 80 :age \"Uh oh\") {}) ; => [:age \"Uh oh\"]
  ```
  "
  ([bound key] (is-less-than bound key (str "must be less than " bound)))
  ([bound key error-msg] (make-claim
                          #(and (contains? %1 key)
                                (pos? (compare bound (get %1 key))))
                          key error-msg)))

(defn is-no-more-than
  "Returns a validator fn that ensures the value of the given field is no more than the value given.

Arguments:
  - `bound` - the maximum permissible value
  - `key` - the key whose value we will check
  - `error-msg` - (optional) custom error message (defaults to \"must be no more than `bound`\")

Return value:
  Returns a validation function that fails if the value of the given field is greater than `bound`.

Example:
  ```
  ((is-no-more-than 80 :age) {:age 80}) ; => nil
  ((is-no-more-than 80 :age) {:age 81}) ; => [:age \"must be no more than 80\"]
  ((is-no-more-than 80 :age \"Uh oh\") {}) ; => [:age \"Uh oh\"]
  ```
  "
  ([bound key] (is-no-more-than bound key (str "must be no more than " bound)))
  ([bound key error-msg] (make-claim
                          #(and (contains? %1 key)
                                (not (neg? (compare bound (get %1 key)))))
                          key error-msg)))
